astropy/cosmology/core.py

```diff
@@ -32,7 +38,265 @@ class CosmologyError(Exception):
+    # -------------------------------------------
+
+    def __call__(self, option, out=None):
+        """Get info on a cosmology class or instance.
+
+        Parameters
+        ----------
+        option : {'parameters', }
+            - 'parameters' : immutable dict of parameter name and value.
+                             If called from a class the value is the Parameter
+                             object itself.
+
+        Returns
+        -------
+        `~types.MappingProxyType` or None
+            MappingProxyType if ``out`` is None and option is "parameters"
+        """
+        out = out if out != "" else sys.stdout
+
+        if option == "parameters":  # works on both class and instance
+            c = c if (c := self._parent) is not None else self._parent_cls
+            info = MappingProxyType({n: getattr(c, n) for n in c.__parameters__})
+
+        else:
+            raise ValueError("'option' must be one of: {'parameters',}")
+
+        return info if out is None else print(info, file=out)
+
+
+class Parameter:
+    """Cosmological parameter (descriptor).
+
+    Should only be used with a :class:`~astropy.cosmology.Cosmology` subclass.
+    For automatic default value and unit inference make sure the Parameter
+    attribute has a corresponding initialization argument (see Examples below).
+
+    Parameters
+    ----------
+    fget : callable or None, optional
+        Function to get the value from instances of the cosmology class.
+        If None (default) returns the corresponding private attribute.
+        Often not set here, but as a decorator with ``getter``.
+    doc : str or None, optional
+        Parameter description. If 'doc' is None and 'fget' is not, then 'doc'
+        is taken from ``fget.__doc__``.
+    unit : unit-like or None (optional, keyword-only)
+        The `~astropy.units.Unit` for the Parameter. If None (default), the
+        unit is inferred from the signature of the containing |Cosmology| class.
+        See :meth:`~astropy.cosmology.Parameter.validate` for usage.
+    equivalencies : `~astropy.units.Equivalency` or sequence thereof
+        Unit equivalencies for this Parameter.
+        See :meth:`~astropy.cosmology.Parameter.validate` for usage.
+    fmt : str (optional, keyword-only)
+        `format` specification, used when making string representation
+        of the containing Cosmology.
+        See https://docs.python.org/3/library/string.html#formatspec
+
+    Examples
+    --------
+    The most common use case of ``Parameter`` is to access the corresponding
+    private attribute.
+
+        >>> from astropy.cosmology import LambdaCDM
+        >>> from astropy.cosmology.core import Parameter
+        >>> class Example1(LambdaCDM):
+        ...     param = Parameter(doc="example parameter")
+        ...     def __init__(self, param=15 * u.m):
+        ...         super().__init__(70, 0.3, 0.7)
+        ...         self._param = Parameter.validate(self.__class__.param, param)
```

Comment:
It would be nice. I tried to have it do that, the problem is ``Ode0``. When this is a Parameter, ``self.Ode0=Ode0`` can route through ``__set__`` and actually be assigned to ``self._Ode0``. However, in flat cosmologies when ``Ode0`` is not a Parameter, but a ``property``, then ``self.Ode0=Ode0`` is wrong (and will raise an AttributeError).

Reply:
My only thought is to have 2 classes: ``Parameter`` and ``NotAParameter`` (with a better name) where ``NotAParameter`` has a ``__set__`` method similar to what you suggest above and I show in https://github.com/astropy/astropy/pull/12136#discussion_r703060118.

Reply:
That might work indeed! But I think it is also OK for the non-parameter ones to just do `self._Ode0 = <calculation>` - in a way, it helps make clear that it is a different beast.

Reply:
But then the non-parameter ones can't call ``super().__init__`` since the base class has ``Ode0`` as a Parameter and will have ``self.Ode0 = ...``, which will raise an `AttributeError`` in this class since ``Ode0`` is no longer a Parameter and is read-only.

Reply:
See https://github.com/astropy/astropy/pull/12136#issuecomment-915618900. I think this problem is tricky and can be tackled at a later date.